File: leetcode/triangle.cpp
#include <string>
#include <math.h>
#include <vector>
using namespace std;
class Solution {
public:
    int minimumTotal(vector<vector<int> > &triangle) {
		vector<vector<int> > tmp=triangle;
		int depth=triangle.size();
		if (depth<=0) return 0;
		for (int i=depth-2;i>=0;i--){
			for (int j=0;j<=i;j++){
				tmp[i][j]=min(tmp[i+1][j],tmp[i+1][j+1])+triangle[i][j];
			}
		}
		return tmp[0][0];
    }
};

File: py/triangle.py
class Solution:
    # @param triangle, a list of lists of integers
    # @return an integer
    def minimumTotal(self, triangle):
        if len(triangle) < 1:
            return
        elif len(triangle) == 1:
            return triangle[0][0]
        
        # len>1
        paths = triangle[0] 
        length = len(triangle)
        for i in range(1, length):
            data = triangle[i]
            size = len(paths)
            temp = [] 
            for j in range(size):
                currentValue = paths[j]
                if j == 0:
                    temp.append(currentValue + data[j])  
                    temp.append(currentValue + data[j + 1]) 
                else:
                    if (currentValue + data[j]) < temp[j]:
                        temp[j] = currentValue + data[j]
                    temp.append(currentValue + data[j + 1])
            paths = temp
        return min(paths)